### PR TITLE
[feat] Add time window for GetTrace in span store interface

### PR DESCRIPTION
**Pull request by rim99**
## Which problem is this PR solving?
Part of https://github.com/jaegertracing/jaeger/issues/4150

## Description of the changes

Refactor trace get method in span store interface:
- use struct instead of literal trace id
- add time window in the new struct

## How was this change tested?
- unit test

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: plugin/storage/grpc/shared/grpc_handler.go [hunk 1]
```diff
@@ -151,7 +151,10 @@ func (s *GRPCHandler) Close(context.Context, *storage_v1.CloseWriterRequest) (*s
 
 // GetTrace takes a traceID and streams a Trace associated with that traceID
 func (s *GRPCHandler) GetTrace(r *storage_v1.GetTraceRequest, stream storage_v1.SpanReaderPlugin_GetTraceServer) error {
-	trace, err := s.impl.SpanReader().GetTrace(stream.Context(), r.TraceID)
+	query := spanstore.GetTraceParameters{
+		TraceID: r.TraceID,
```
Comment (yurishkuro) [inline plugin/storage/grpc/shared/grpc_handler.go:155]: You are doing the PRs in the wrong order. I recommend first starting with upgrading grpc storage API, then here you'd be able to assign time stamps right away, instead of leaving unintentional bug.

File: plugin/storage/grpc/shared/grpc_client_test.go [hunk 2]
```diff
@@ -406,7 +412,9 @@ func TestGrpcClientGetArchiveTrace_NoTrace(t *testing.T) {
 			TraceID: mockTraceID,
 		}).Return(nil, spanstore.ErrTraceNotFound)
 
-		s, err := r.client.ArchiveSpanReader().GetTrace(context.Background(), mockTraceID)
+		s, err := r.client.ArchiveSpanReader().GetTrace(context.Background(), spanstore.GetTraceParameters{
+			TraceID: mockTraceID,
+		})
```
Comment (yurishkuro) [inline plugin/storage/grpc/shared/grpc_client_test.go:417]:
```suggestion
		s, err := r.client.ArchiveSpanReader().GetTrace(
			context.Background(),
			spanstore.GetTraceParameters{
				TraceID: mockTraceID,
			},
		)
```

File: plugin/storage/es/spanstore/reader.go [hunk 3]
```diff
@@ -233,11 +233,11 @@ func timeRangeIndices(indexName, indexDateLayout string, startTime time.Time, en
 }
 
 // GetTrace takes a traceID and returns a Trace associated with that traceID
-func (s *SpanReader) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+func (s *SpanReader) GetTrace(ctx context.Context, query spanstore.GetTraceParameters) (*model.Trace, error) {
 	ctx, span := s.tracer.Start(ctx, "GetTrace")
 	defer span.End()
 	currentTime := time.Now()
-	traces, err := s.multiRead(ctx, []model.TraceID{traceID}, currentTime.Add(-s.maxSpanAge), currentTime)
+	traces, err := s.multiRead(ctx, []model.TraceID{query.TraceID}, currentTime.Add(-s.maxSpanAge), currentTime)
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/reader.go:240]: add TODO to use timestamps, they can help limit the number of different indices that must be inspected

File: cmd/query/app/http_handler.go [hunk 4]
```diff
@@ -266,7 +266,10 @@ func (aH *APIHandler) tracesByIDs(ctx context.Context, traceIDs []model.TraceID)
 	var traceErrors []structuredError
 	retMe := make([]*model.Trace, 0, len(traceIDs))
 	for _, traceID := range traceIDs {
-		if trc, err := aH.queryService.GetTrace(ctx, traceID); err != nil {
+		query := spanstore.GetTraceParameters{
```
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:269]: somewhere here the timestamps must be handled, so leave a TODO. This could be another PR to precede this one, where we introduce timestamps params in /api/trace/{traceID} API.

File: cmd/query/app/grpc_handler.go [hunk 5]
```diff
@@ -89,7 +89,10 @@ func (g *GRPCHandler) GetTrace(r *api_v2.GetTraceRequest, stream api_v2.QuerySer
 	if r.TraceID == (model.TraceID{}) {
 		return errUninitializedTraceID
 	}
-	trace, err := g.queryService.GetTrace(stream.Context(), r.TraceID)
+	query := spanstore.GetTraceParameters{
+		TraceID: r.TraceID,
```
Comment (yurishkuro) [inline cmd/query/app/grpc_handler.go:93]: copy timestamps from request

File: cmd/query/app/grpc_handler.go [hunk 6]
```diff
@@ -109,7 +112,10 @@ func (g *GRPCHandler) ArchiveTrace(ctx context.Context, r *api_v2.ArchiveTraceRe
 	if r.TraceID == (model.TraceID{}) {
 		return nil, errUninitializedTraceID
 	}
-	err := g.queryService.ArchiveTrace(ctx, r.TraceID)
+	query := spanstore.GetTraceParameters{
+		TraceID: r.TraceID,
```
Comment (yurishkuro) [inline cmd/query/app/grpc_handler.go:116]: copy timestamps from request

File: cmd/query/app/apiv3/http_gateway.go [hunk 7]
```diff
@@ -136,7 +136,10 @@ func (h *HTTPGateway) getTrace(w http.ResponseWriter, r *http.Request) {
 	if h.tryParamError(w, err, paramTraceID) {
 		return
 	}
-	trc, err := h.QueryService.GetTrace(r.Context(), traceID)
+	query := spanstore.GetTraceParameters{
+		TraceID: traceID,
```
Comment (yurishkuro) [inline cmd/query/app/apiv3/http_gateway.go:140]: copy timestamps from request

File: cmd/query/app/apiv3/grpc_handler.go [hunk 8]
```diff
@@ -31,8 +31,11 @@ func (h *Handler) GetTrace(request *api_v3.GetTraceRequest, stream api_v3.QueryS
 	if err != nil {
 		return fmt.Errorf("malform trace ID: %w", err)
 	}
+	query := spanstore.GetTraceParameters{
+		TraceID: traceID,
```
Comment (yurishkuro) [inline cmd/query/app/apiv3/grpc_handler.go:35]: copy timestamps from request

File: cmd/query/app/apiv3/gateway_test.go [hunk 9]
```diff
@@ -140,18 +141,18 @@ func (gw *testGateway) runGatewayGetOperations(t *testing.T) {
 }
 
 func (gw *testGateway) runGatewayGetTrace(t *testing.T) {
-	trace, traceID := makeTestTrace()
-	gw.reader.On("GetTrace", matchContext, traceID).Return(trace, nil).Once()
-	gw.getTracesAndVerify(t, "/api/v3/traces/"+traceID.String(), traceID)
+	trace, query := makeTestTrace()
+	gw.reader.On("GetTrace", matchContext, query).Return(trace, nil).Once()
+	gw.getTracesAndVerify(t, "/api/v3/traces/"+query.TraceID.String(), query.TraceID)
```
Comment (yurishkuro) [inline cmd/query/app/apiv3/gateway_test.go:146]: could introduce a preliminary PR that adds timestamps args to the REST API calls

File: cmd/jaeger/internal/integration/span_reader.go [hunk 10]
```diff
@@ -66,9 +66,9 @@ func unwrapNotFoundErr(err error) error {
 	return err
 }
 
-func (r *spanReader) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+func (r *spanReader) GetTrace(ctx context.Context, query spanstore.GetTraceParameters) (*model.Trace, error) {
 	stream, err := r.client.GetTrace(ctx, &api_v2.GetTraceRequest{
-		TraceID: traceID,
+		TraceID: query.TraceID,
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/span_reader.go:71]: copy timestamps from request

File: plugin/storage/grpc/shared/grpc_handler.go (hunk 1 above)
Comment (rim99) [reply plugin/storage/grpc/shared/grpc_handler.go:155]: Got your point, let's try this first: https://github.com/jaegertracing/jaeger/pull/6244

File: cmd/query/app/apiv3/http_gateway.go (hunk 7 above)
Comment (rim99) [reply cmd/query/app/apiv3/http_gateway.go:143]: Added TODO, since API changes will be in another PR

File: cmd/query/app/apiv3/gateway_test.go (hunk 9 above)
Comment (rim99) [reply cmd/query/app/apiv3/gateway_test.go:146]: here: https://github.com/jaegertracing/jaeger/pull/6248

File: cmd/query/app/apiv3/grpc_handler_test.go [hunk 11]
```diff
@@ -79,7 +79,50 @@ func newTestServerClient(t *testing.T) *testServerClient {
...
-	tsc.reader.On("GetTrace", matchContext, matchTraceID).Return(
+	traceIdLiteral := "156"
+	traceId, _ := model.TraceIDFromString(traceIdLiteral)
+
+	expectedTraceGetParameters := spanstore.GetTraceParameters{
+		TraceID:   traceId,
+		StartTime: time.Time{},
+		EndTime:   time.Time{},
+	}
+	tsc.reader.On("GetTrace", matchContext, expectedTraceGetParameters).Return(
```
Comment (yurishkuro) [inline cmd/query/app/apiv3/grpc_handler_test.go:90]: too many redundant code here. You can make the existing test to be table driven

File: cmd/query/app/grpc_handler.go [hunk 12]
```diff
@@ -109,7 +114,16 @@ func (g *GRPCHandler) ArchiveTrace(ctx context.Context, r *api_v2.ArchiveTraceRe
...
+	}
+	if r.StartTime != nil {
```
Comment (yurishkuro) [inline cmd/query/app/grpc_handler.go:120]: inconsistent with other requests - do we need to regenerate ArchiveTraceRequest with nullable=false?

File: cmd/query/app/http_handler.go (hunk 4 above)
Comment (yurishkuro) [reply cmd/query/app/http_handler.go:269]: how is this comment resolved?

File: storage/spanstore/interface.go [hunk 13]
```diff
@@ -50,6 +50,13 @@ type Reader interface {
 	FindTraceIDs(ctx context.Context, query *TraceQueryParameters) ([]model.TraceID, error)
 }
 
+// GetTraceParameters contains parameters of a trace get.
+type GetTraceParameters struct {
+	TraceID   model.TraceID
+	StartTime time.Time
+	EndTime   time.Time
```
Comment (yurishkuro) [inline storage/spanstore/interface.go:57]:
```suggestion
	StartTime time.Time // optional
	EndTime   time.Time // optional
```

File: storage/spanstore/metrics/decorator.go [hunk 14]
```diff
@@ -78,9 +78,9 @@ func (m *ReadMetricsDecorator) FindTraceIDs(ctx context.Context, traceQuery *spa
 }
 
 // GetTrace implements spanstore.Reader#GetTrace
-func (m *ReadMetricsDecorator) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+func (m *ReadMetricsDecorator) GetTrace(ctx context.Context, traceGet spanstore.GetTraceParameters) (*model.Trace, error) {
```
Comment (yurishkuro) [inline storage/spanstore/metrics/decorator.go:81]:
```suggestion
func (m *ReadMetricsDecorator) GetTrace(ctx context.Context, query spanstore.GetTraceParameters) (*model.Trace, error) {
```

File: storage/spanstore/metrics/decorator_test.go [hunk 15]
```diff
@@ -29,8 +29,8 @@ func TestSuccessfulUnderlyingCalls(t *testing.T) {
 	mockReader.On("GetOperations", context.Background(), operationQuery).
 		Return([]spanstore.Operation{}, nil)
 	mrs.GetOperations(context.Background(), operationQuery)
-	mockReader.On("GetTrace", context.Background(), model.TraceID{}).Return(&model.Trace{}, nil)
-	mrs.GetTrace(context.Background(), model.TraceID{})
+	mockReader.On("GetTrace", context.Background(), spanstore.GetTraceParameters{TraceID: model.TraceID{}}).Return(&model.Trace{}, nil)
+	mrs.GetTrace(context.Background(), spanstore.GetTraceParameters{TraceID: model.TraceID{}})
```
Comment (yurishkuro) [inline storage/spanstore/metrics/decorator_test.go:33]:
```suggestion
	mrs.GetTrace(context.Background(), spanstore.GetTraceParameters{})
```

File: cmd/query/app/apiv3/grpc_handler_test.go (hunk 11 above)
Comment (rim99) [reply cmd/query/app/apiv3/grpc_handler_test.go:90]: done

File: cmd/query/app/http_handler.go (hunk 4 above)
Comment (rim99) [reply cmd/query/app/http_handler.go:270]: Added more TODOs

**Comment by rim99** [comment]
PR updated, please review, thanks @yurishkuro

File: cmd/query/app/apiv3/gateway_test.go (hunk 9 above)
Comment (yurishkuro) [reply cmd/query/app/apiv3/gateway_test.go:146]: that PR doesn't seem to compile, can you get it to green state?

File: cmd/query/app/apiv3/grpc_handler_test.go [hunk 16]
```diff
@@ -78,33 +78,59 @@ func newTestServerClient(t *testing.T) *testServerClient {
...
+			api_v3.GetTraceRequest{
+				TraceId:   traceIdLiteral,
+				StartTime: time.Unix(1, 2).UTC(),
+				EndTime:   time.Unix(3, 4).UTC(),
+			},
+		},
+	}
+
+	for _, input := range inputs {
+		tsc := newTestServerClient(t)
```
Comment (yurishkuro) [inline cmd/query/app/apiv3/grpc_handler_test.go:110]:
```suggestion
	for _, input := range inputs {
		t.Run(input.name, func(t *testing.T) {
			tsc := newTestServerClient(t)
			...
		})
```

Comment (rim99) [reply cmd/query/app/apiv3/grpc_handler_test.go]: done

**Comment by yurishkuro** [comment]
@rim99 could you please resolve the conflicts?

**Comment by rim99** [comment]
> @rim99 could you please resolve the conflicts?

Resolved. Could you please reivew?

**Comment by yurishkuro** [comment]
TODOs added to 
  * anonymizer/app/query/query.go
  * cmd/query/app/apiv3/http_gateway.go
  * cmd/query/app/http_handler.go (3 times)
  * plugin/storage/es/spanstore/reader.go

**Review by yurishkuro — APPROVED** [review]
LGTM
